### PR TITLE
Base64 encoding/decoding using SIMDe

### PR DESCRIPTION
**Pull request by cbielow**
## Description

faster de&encoding of any XML data which contains base64 strings, e.g. mzML.
FileConverter is about 8-10% faster when loading+storing an mzML file on both Linux and Windows.

We now require some advanced instruction sets for x64 CPUs: SSE3 (g++/clang) or AVX (MSVC compiler for Windows); and NEON for ARM64 CPUs.
AVX was introduced 2013 - we may exclude some really old CPU's for Windows here, but everybody else should see significant benefits.
SSE3 was introduced 2004 -- not a problem here.

## Checklist
- [ ] Make sure that you are listed in the AUTHORS file
- [ ] Add relevant changes and new features to the CHANGELOG file
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] New and existing unit tests pass locally with my changes
- [ ] Updated or added python bindings for changed or new classes (Tick if no updates were necessary.)

### How can I get additional information on failed tests during CI
<details>
  <summary>Click to expand</summary>
If your PR is failing you can check out

- The details of the action statuses at the end of the PR or the "Checks" tab.
- http://cdash.openms.de/index.php?project=OpenMS and look for your PR. Use the "Show filters" capability on the top right to search for your PR number.
  If you click in the column that lists the failed tests you will get detailed error messages.

</details>

### Advanced commands (admins / reviewer only)
<details>
  <summary>Click to expand</summary>
  
- `/reformat` (experimental) applies the clang-format style changes as additional commit. Note: your branch must have a different name (e.g., yourrepo:feature/XYZ) than the receiving branch (e.g., OpenMS:develop). Otherwise, reformat fails to push.
- setting the label "NoJenkins" will skip tests for this PR on jenkins (saves resources e.g., on edits that do not affect tests)
- commenting with `rebuild jenkins` will retrigger Jenkins-based CI builds
  
</details>

---
:warning: Note: Once you opened a PR try to minimize the number of *pushes* to it as every push will trigger CI (automated builds and test) and is rather heavy on our infrastructure (e.g., if several pushes per day are performed).

### CODE REVIEW

**Review by jpfeuffer — COMMENTED** [review]
General comment before looking at the code:

How bad would it be to have this as a  build option? I'm not sure about all its implications yet.

Does this allow linking to an existing simde lib installation? It's on all major package managers as far as I can see.

Also, if we make it default it would make sense to adapt the contrib to build everything (especially Eigen) with that support (as they will benefit as well).

**Comment by cbielow** [comment]
> How bad would it be to have this as a build option? I'm not sure about all its implications yet.

it would mean that we need to keep two implementations around (and make sure we test both of them), i.e. this would double CI costs, if done properly.
Any specific things we should be concerned about when switching to SSE/AVX?

 
> Does this allow linking to an existing simde lib installation? It's on all major package managers as far as I can see.

Good point! I will add that.

> 
> Also, if we make it default it would make sense to adapt the contrib to build everything (especially Eigen) with that support (as they will benefit as well).

This probably needs careful consideration for each individual library, since this may have implications on their tests and we may need to pass these things into the respective build systems. But that can be done, yes.

**Comment by jpfeuffer** [comment]
The reason is mainly conda:
https://longr.github.io/posts/gromacs-conda/

**Comment by timosachsenberg** [comment]
I am not sure if I understand all implications.
Some technical questions
- Would it make sense to use the single header version with https://github.com/simd-everywhere/simde/blob/master/amalgamate.py - not sure how often we would need to update it if we use only the older instruction sets
- Could the simd stuff be put into the cpp files not exposing SIMD specific types or functions?
- Eigen is header only. Can we also enable it there?

**Comment by cbielow** [comment]
> I am not sure if I understand all implications. Some technical questions
> 
>     * Would it make sense to use the single header version with https://github.com/simd-everywhere/simde/blob/master/amalgamate.py - not sure how often we would need to update it if we use only the older instruction sets

Since we will have to allow for an external SIMDe anyway, I don't think there is an advantage. We'd even have to conditionally change the way we include SIMDe headers, since an external SIMDe will use different names then.

> 
>     * Could the simd stuff be put into the cpp files not exposing SIMD specific types or functions?

This may work for Base64. Not sure about other potential applications. I can see why hiding the SIMDe headers from the OpenMS interface is beneficial, but if we allow for a custom external SIMDe, the problem is somewhat reduced.
So I'm not sure if its worth the effort.

> 
>     * Eigen is header only. Can we also enable it there?

Yes, that's easy. Just use the compiler flags, as we already do:
https://eigen.tuxfamily.org/index.php?title=FAQ#Which_SIMD_instruction_sets_are_supported_by_Eigen.3F

**Comment by cbielow** [comment]
> The reason is mainly conda: https://longr.github.io/posts/gromacs-conda/

I've read some conda discussions and we seem to get away with SSE3 on x64: https://github.com/conda-forge/conda-forge.github.io/issues/49#issuecomment-383768643
We are currently not building for Windows, right? So there should be no problem.

**Comment by timosachsenberg** [comment]
other questions: OpenMP4 SIMD support came to visual studio 2019 https://learn.microsoft.com/en-us/cpp/build/reference/openmp-enable-openmp-2-0-support?view=msvc-170
others have it for a long time. I guess this was not an option?

**Comment by cbielow** [comment]
AFAIK OpenMP simd is only for loops. This is not enough for Base64 encoding/decoding.

**Comment by poshul** [comment]
>  AVX was introduced 2013 - we may exclude some really old CPU's for Windows here, but everybody else should see significant benefits.

AVX was introduced in 2011 (AVX2 was 2013). So the user would have to have something Westmere/Nehalem or older for this to be an issue. Given that Microsoft didn't even support these processors in Windows 8.1 I don't feel bad about any compatibility issues there.

More generally are there likely to be other parts of our code that we could improve with SIMDe?

**Comment by timosachsenberg** [comment]
I would not call it "improve" because there are pros (mainly a moderate performance gain)/cons associated with this. But my personal impression (up to this point) is that the applicability of the library is a bit limited if we can use OpenMP 4 SIMD instructions with loops already.

**Comment by cbielow** [comment]
It depends on what you regard as 'moderate'.
For individual functions you can expect somewhere between 3-10x for SSE (Base64 is about 5x, and the 'removeWhitespace' function can be 10x faster (PR coming up).
There are plenty of other places where SSE/AVX are useful, e.g. Eigen3.

**Comment by timosachsenberg** [comment]
ah ok thanks for the clarification. Pitty that this only translates to an overall speedup of 8%-10%

**Comment by cbielow** [comment]
Well, you cannot do 'em all at once.
Once `removeWhitespaces` is SIMD-capable, the speedup will be 20% for fileconverter...

BTW: also RedHat mandates x86-64-v2 (roughly SSE4) as of last year for RedHat9 (https://developers.redhat.com/blog/2021/01/05/building-red-hat-enterprise-linux-9-for-the-x86-64-v2-microarchitecture-level#)

**Comment by timosachsenberg** [comment]
don't get me wrong. I am all in for moving to SSE3+ if that doesn't break packages. I just want to understand if the extra dependency is worth it.

File: src/openms/thirdparty/simde/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+# OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF
+# ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+#
+# --------------------------------------------------------------------------
+# $Maintainer: Chris Bielow $
+# $Authors: Chris Bielow $
+# --------------------------------------------------------------------------
+
+project("SIMDe")
+set(SIMDE_INCLUDE_DIRECTORY "${PROJECT_SOURCE_DIR}" CACHE PATH "Directory for SIMDe headers")
```
Comment (jpfeuffer) [inline src/openms/thirdparty/simde/CMakeLists.txt]: if you create a target here (as in your find module), you can use the more modern target_link_library approach.
For header-only libs not much advantages but maybe we manage to do everything with targets at some point and have it unified.

**Comment by jpfeuffer** [comment]
If you could confirm with the conda and debian guys that this is ok, I would be willing to merge if it passes.

**Comment by cbielow** [comment]
rebuild jenkins

**Comment by cbielow** [comment]
rebuild jenkins

**Comment by cbielow** [comment]
rebuild jenkins

**Comment by cbielow** [comment]
this is ready for merge. Please have a final look :)

**Comment by cbielow** [comment]
btw: Debian devs said, we should just mention SSE in Changelog/Readme. That's enough for package creation.

**Comment by cbielow** [comment]
tests pass

**Review by timosachsenberg — APPROVED** [review]

**Pull request merged**